filter.go

```diff
@@ -215,50 +217,63 @@ var filterBufPool = sync.Pool{
 	},
 }
 
-func releaseRegexpFilterBuf(l []Line) {
+func releaseFilterLineBuf(l []Line) {
 	if l == nil {
 		return
 	}
 	l = l[0:0]
 	filterBufPool.Put(l)
 }
 
-func getRegexpFilterBuf() []Line {
+func getFilterLineBuf() []Line {
 	l := filterBufPool.Get().([]Line)
 	return l
 }
 
+type filter interface {
+	filter(Line) (Line, error)
+}
+
+// This flusher is run in a separate goroutine so that the filter can
+// run separately from accepting incoming messages
+func flusher(f filter, incoming chan []Line, done chan struct{}, out pipeline.OutputChannel) {
+	if pdebug.Enabled {
+		g := pdebug.Marker("flusher goroutine")
+		defer g.End()
+	}
+
+	defer close(done)
```

Comment:
this one is required. we need to wait until all output has been sent through `out`